pkg/controller/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,318 @@
+			c.mu.RUnlock()
+
+			c.makeUpdatesForSpecInTime(ctx, currentGVKsInAgg, excluderChanged)
+		}
+	}
+}
+
+// listAndSyncData returns a set of gvks that were successfully listed and synced.
+func (c *CacheManager) listAndSyncData(ctx context.Context, gvks []schema.GroupVersionKind, reader client.Reader) *watch.Set {
+	gvksSuccessfullySynced := watch.NewSet()
+	for _, gvk := range gvks {
+		err := c.listAndSyncDataForGVK(ctx, gvk)
+		if err != nil {
+			log.Error(err, "internal: error syncing gvks cache data")
+			// we don't remove this gvk as we will try to re-add it later
+			// we also don't return on this error to be able to list and sync
+			// other gvks in order to protect against a bad gvk.
+		} else {
+			gvksSuccessfullySynced.Add(gvk)
+		}
+	}
+	return gvksSuccessfullySynced
+}
+
+// makeUpdatesForSpecInTime performs a conditional wipe followed by a replay if necessary as
+// given by the current spec (currentGVKsInAgg, excluderChanged) at the time of the call.
+func (c *CacheManager) makeUpdatesForSpecInTime(ctx context.Context, currentGVKsInAgg *watch.Set, excluderChanged bool) {
+	if c.watchedSet.Equals(currentGVKsInAgg) && !excluderChanged {
+		return // nothing to do if both sets are the same and the excluder didn't change
+	}
+
+	// replace the current watch set for the sync_controller to pick up
+	// any updates on said GVKs.
+	// also save the current watch set to make cache changes later
+	oldWatchSet := watch.NewSet()
+	oldWatchSet.AddSet(c.watchedSet)
+
+	var innerError error
+	c.watchedSet.Replace(currentGVKsInAgg, func() {
+		// *Note the following steps are not transactional with respect to admission control*
+
+		// Important: dynamic watches update must happen *after* updating our watchSet.
+		// Otherwise, the sync controller will drop events for the newly watched kinds.
+		// Defer error handling so object re-sync happens even if the watch is hard
+		// errored due to a missing GVK in the watch set.
+		innerError = c.registrar.ReplaceWatch(ctx, currentGVKsInAgg.Items())
+	})
+	if innerError != nil {
+		log.Error(innerError, "internal: error replacing watch set")
+	}
+
+	gvksToDelete := oldWatchSet.Difference(currentGVKsInAgg).Items()
+	newGVKsToSync := currentGVKsInAgg.Difference(oldWatchSet)
+
+	// remove any gvks not needing to be synced anymore
+	// or re evaluate all if the excluder changed.
+	if len(gvksToDelete) > 0 || excluderChanged {
+		if err := c.wipeData(ctx); err != nil {
+			log.Error(err, "internal: error wiping cache")
+		}
+
+		if excluderChanged {
+			c.unsetExcluderChanged()
+		}
+
+		// everything that gets wiped needs to be readded
+		newGVKsToSync.AddSet(currentGVKsInAgg)
+	}
+
+	// sync net new gvks and potentially replayed gvks from the cache wipe above
+	gvksSynced := c.listAndSyncData(ctx, newGVKsToSync.Items(), c.reader)
+
+	gvksNotSynced := gvksSynced.Difference(newGVKsToSync)
+	for _, gvk := range gvksNotSynced.Items() {
+		log.Info(fmt.Sprintf("failed to sync gvk: %s; will retry", gvk))
```

Comment:
What logic initiates the retry?

Reply:
thanks for this call out! as of now, there is no retry for these gvks. yes the ticker will tick but if the watched set is the same as the gvk aggregator at the time of `makeUpdates`, then, currently, there won't be a way to retry these. 

I have an idea to "remember" these so we can indeed retry the listing for them.

Reply:
retry moved in a separate loop